blob/s3blob/s3blob.go

```diff
@@ -120,32 +125,82 @@ const Scheme = "s3"
 
 // OpenBucketURL opens the S3 bucket with the same name as the host in the URL.
 func (o *URLOpener) OpenBucketURL(ctx context.Context, u *url.URL) (*blob.Bucket, error) {
-	for k := range u.Query() {
-		return nil, fmt.Errorf("open S3 bucket %q: unknown query parameter %s", u, k)
+	opts, err := o.forParams(ctx, u.Query())
+	if err != nil {
+		return nil, fmt.Errorf("open bucket %v: %v", u, err)
+	}
+	return OpenBucket(ctx, o.ConfigProvider, u.Host, opts)
+}
+
+var legalQueryParam = map[string]bool{
+	"region":           true,
+	"endpoint":         true,
+	"disableSSL":       true,
+	"s3ForcePathStyle": true,
+}
+
+func (o *URLOpener) forParams(ctx context.Context, q url.Values) (*Options, error) {
+	for k := range q {
+		if !legalQueryParam[k] {
+			return nil, fmt.Errorf("unknown S3 query parameter %s", k)
+		}
 	}
-	return OpenBucket(ctx, o.ConfigProvider, u.Host, &o.Options)
+	var cfg aws.Config
+	override := false
+	if region := q["region"]; len(region) > 0 {
+		cfg.Region = aws.String(region[0])
+		override = true
+	}
+	if endpoint := q["endpoint"]; len(endpoint) > 0 {
+		cfg.Endpoint = aws.String(endpoint[0])
+		override = true
+	}
+	if disableSSL := q["disableSSL"]; len(disableSSL) > 0 {
+		cfg.DisableSSL = aws.Bool(disableSSL[0] == "true")
+		override = true
+	}
+	if s3ForcePathStyle := q["s3ForcePathStyle"]; len(s3ForcePathStyle) > 0 {
+		cfg.S3ForcePathStyle = aws.Bool(s3ForcePathStyle[0] == "true")
+		override = true
+	}
+	if !override {
+		return &o.Options, nil
+	}
+	opts := new(Options)
+	*opts = o.Options
+	opts.Cfgs = append(opts.Cfgs, &cfg)
+	return opts, nil
 }
 
 // Options sets options for constructing a *blob.Bucket backed by fileblob.
-type Options struct{}
+type Options struct {
+	// Cfgs holds configuration overrides for the S3 client. For example, you
+	// may want to use the same *session.Session for buckets across multiple
+	// regions; you can override the aws.Config.Region here.
+	Cfgs []*aws.Config
```

Comment:
I don't think adding an option is necessary. See my diff https://github.com/google/go-cloud/compare/master...zombiezen:s3-region-url for a variant that simplifies the logic here in a more generalized way.

Reply:
Cool; done.